src/swap/utils.ts

```diff
@@ -1,243 +1,250 @@
+    const quoteGetterArgs = {
+      pool: v1_1Pool.info,
+      assetOut: {amount, id: Number(assetOut.id)},
+      decimals: {assetIn: assetIn.decimals, assetOut: assetOut.decimals},
+      reserves: v1_1Pool.reserves
+    };
+
+    quotePromises.push(
+      new Promise((resolve, reject) => {
+        try {
+          resolve(SwapV1_1.getFixedOutputSwapQuote(quoteGetterArgs));
+        } catch (error) {
+          reject(error);
         }
+      })
+    );
+  } else {
+    quotePromises.push(
+      Promise.reject(
+        new SwapQuoteError(
+          SwapQuoteErrorType.NoAvailablePoolError,
+          "Trying to swap from non-existent pool"
+        )
+      )
+    );
+  }
 
-        resolve({pool, quote});
-      } catch (error) {
-        reject(error);
-      }
-    });
-  });
+  const v2Pool = pools.find((pool) => pool.info.contractVersion === CONTRACT_VERSION.V2);
+
+  quotePromises.push(
+    SwapV2.getFixedOutputSwapQuote({
+      amount,
+      assetInID: getAssetId(assetIn),
+      assetOutID: getAssetId(assetOut),
+      pool: v2Pool?.info ?? null,
+      decimals: {assetIn: assetIn.decimals, assetOut: assetOut.decimals},
+      isSwapRouterEnabled,
+      network: params.network
+    })
+  );
 
   const validQuotes = await validateQuotes(quotePromises);
 
   return getBestQuote(validQuotes);
 }
 
-/**
- * Compares the given quotes and returns the best one (with the highest rate).
- */
-function getBestQuote(quotes: SwapQuoteWithPool[]): SwapQuoteWithPool {
-  const quotesByDescendingRate = quotes
-    .filter((quote) => !isPoolEmpty(quote.pool.reserves))
-    .sort((a, b) => b.quote.rate - a.quote.rate);
-
-  return quotesByDescendingRate[0];
-}
-
-export function generateTxns(params: {
-  client: Algodv2;
-  pool: V1PoolInfo | V2PoolInfo;
-  poolAddress: string;
-  swapType: SwapType;
-  assetIn: AssetWithIdAndAmount;
-  assetOut: AssetWithIdAndAmount;
-  slippage: number;
-  initiatorAddr: string;
-}): Promise<SignerTransaction[]> {
-  if (params.pool.contractVersion === CONTRACT_VERSION.V1_1) {
-    return SwapV1_1.generateTxns(params);
+export function generateTxns(
+  params: GenerateSwapTxnsParams
+): Promise<SignerTransaction[]> {
+  if (
+    params.quote.type === SwapQuoteType.Direct &&
+    getSwapQuoteContractVersion(params.quote) === CONTRACT_VERSION.V1_1
+  ) {
+    return SwapV1_1.generateTxns({...params, quote: params.quote.data});
   }
 
   return SwapV2.generateTxns(params);
 }
 
 export function signTxns(params: {
-  pool: V1PoolInfo;
+  quote: SwapQuote;
   txGroup: SignerTransaction[];
   initiatorSigner: InitiatorSigner;
 }): Promise<Uint8Array[]> {
-  if (params.pool.contractVersion === CONTRACT_VERSION.V1_1) {
-    return SwapV1_1.signTxns(params);
+  if (
+    params.quote.type === SwapQuoteType.Direct &&
+    getSwapQuoteContractVersion(params.quote) === CONTRACT_VERSION.V1_1
+  ) {
+    const {
+      data: {pool}
+    } = params.quote;
+
+    return SwapV1_1.signTxns({pool, ...params});
```

Comment:
it may be better to always override the pool object that may come from params. It is highly unlikely if they are using TS but still

```suggestion
    return SwapV1_1.signTxns({...params, pool});
```